src/dyngen.cpp

```diff
@@ -158,55 +208,86 @@ void DynGen::updateCode(const std::string &code) {
 // the library by traversing the library which is a linked list.
 // If the hash ID already exists the code gets updated and all running
 // instances should be updated.
-bool enterFileToDynGenLibrary(World *world, void *raw_callback) {
-  auto entry = static_cast<NewDynGenLibraryEntry*>(raw_callback);
+bool loadFileToDynGenLibrary(World *world, void *rawCallbackData) {
+  auto entry = static_cast<NewDynGenLibraryEntry*>(rawCallbackData);
+
+  auto codeFile = std::ifstream(entry->codePath, std::ios::binary);
+  if (!codeFile.is_open()) {
+    std::cerr << "Could not open DynGen file at " << entry->codePath << std::endl;
+    return false;
+  }
 
-  // read file, see https://stackoverflow.com/a/19922123
-  std::ifstream codeFile;
-  codeFile.open(entry->codePath);
   std::stringstream codeStream;
-  codeStream << codeFile.rdbuf();
-  const std::string code = codeStream.str();
+
+  codeFile.seekg(0, std::ios::end);
+  const std::streamsize codeSize = codeFile.tellg();
+  codeFile.seekg(0);
+
+  // add /0
+  auto* codeBuffer = new char[codeSize + 1];
+  codeFile.read(codeBuffer, codeSize);
+  codeBuffer[codeSize] = '\0';
+
+  entry->code = codeBuffer;
+
+  // continue to next stage
+  return true;
+}
+
+// runs in stage 3 (RT-thread)
+bool swapCode(World* world, void *rawCallbackData) {
+  auto entry = static_cast<NewDynGenLibraryEntry*>(rawCallbackData);
 
   CodeLibrary* node = gLibrary;
-  CodeLibrary* prevNode = nullptr;
   while (node && node->id != entry->hash) {
-    prevNode = node;
     node = node->next;
   }
 
   if (!node) {
-    auto* newNode = new CodeLibrary {
-      gLibrary,
-      entry->hash,
-      nullptr,
-      code,
-    };
+    auto* newNode = static_cast<CodeLibrary*>(RTAlloc(world, sizeof(CodeLibrary)));
+    newNode->next = gLibrary;
+    newNode->id = entry->hash;
+    newNode->dynGenNodes = nullptr;
+    newNode->code = entry->code;
     gLibrary = newNode;
   } else {
-    node->code = code;
+    // swap code
+    entry->oldCode = node->code;
+    node->code = entry->code;
     for (auto* unit = node->dynGenNodes; unit; unit = unit->next) {
-      unit->dynGenUnit->updateCode(code);
+      // protecting in case the vm already got removed b/c the synth got removed
+      if (unit->dynGenUnit->mVm != nullptr) {
+        unit->dynGenUnit->mVm->init(entry->code);
```

Comment:
You can't init the VM on the RT thread! Instead add an `updateCode` method to `DynGen` that internally spawns an async command, just like you do in the constructor. In fact, you can probably share quite a bit of code between these two.